File: scripts/scaffold_component.py
#!/usr/bin/env python3
"""
组件脚手架生成器 - 基于系统预制模板

从 .ai-workflow/templates/component/ 拷贝干净的标准模板，替换占位符。

这是创建新组件的推荐方式，提供最小化、标准化的组件结构，
保证可编译运行，适合所有组件类型。
"""

import argparse
import json
import sys
import shutil
import re
from pathlib import Path
from datetime import datetime
from typing import Dict

# 项目根目录
PROJECT_ROOT = Path(__file__).parent.parent
TEMPLATE_DIR = PROJECT_ROOT / ".ai-workflow" / "templates" / "component"


class ComponentScaffolder:
    """组件脚手架生成器"""
    
    def __init__(self, component: str, mode: str, template_data: Dict):
        self.component = component
        self.mode = mode
        self.data = template_data
        self.component_dir = PROJECT_ROOT / "components" / component
        
        # 生成占位符映射
        self.placeholders = self._generate_placeholders()
        
    def _generate_placeholders(self) -> Dict[str, str]:
        """生成占位符映射"""
        # 将 pep-my-component 转换为 PepMyComponent
        parts = self.component.split('-')
        pascal_case = ''.join(word.capitalize() for word in parts)
        
        return {
            '{{COMPONENT_NAME}}': self.component,
            '{{COMPONENT_NAME_PASCAL}}': pascal_case,
            '{{DESCRIPTION}}': self.data.get('description', f'{self.component} 组件'),
            '{{INIT_DATE}}': datetime.now().isoformat(),
        }
        
    def generate(self):
        """生成脚手架"""
        print(f"🏗️  从模板创建 {self.component}")
        
        # 检查模板是否存在
        if not TEMPLATE_DIR.exists():
            print(f"❌ 模板目录不存在: {TEMPLATE_DIR}")
            print(f"请确保 .ai-workflow/templates/component/ 目录存在")
            sys.exit(1)
        
        # 检查组件是否已存在
        if self.component_dir.exists():
            print(f"❌ 组件已存在: {self.component}")
            sys.exit(1)
        
        # 1. 拷贝模板目录
        print(f"📦 拷贝模板...")
        shutil.copytree(TEMPLATE_DIR, self.component_dir)
        
        # 2. 替换所有文件中的占位符
        print(f"✏️  替换占位符...")
        self._replace_placeholders_in_directory(self.component_dir)
        
        # 3. 重命名包含占位符的文件
        print(f"📝 重命名文件...")
        self._rename_placeholder_files(self.component_dir)
        
        # 4. 根据 mode 进行增强（如果需要）
        if self.mode == 'custom' and self.data:
            print(f"🎨 应用定制增强...")
            self._apply_custom_enhancements()
        
        print(f"✅ 组件脚手架创建完成")
        print(f"📂 位置: components/{self.component}")
        
    def _replace_placeholders_in_directory(self, directory: Path):
        """递归替换目录中所有文件的占位符"""
        for item in directory.rglob('*'):
            if item.is_file():
                self._replace_placeholders_in_file(item)
                
    def _replace_placeholders_in_file(self, file_path: Path):
        """替换单个文件中的占位符"""
        try:
            # 读取文件内容
            content = file_path.read_text(encoding='utf-8')
            
            # 替换所有占位符
            for placeholder, value in self.placeholders.items():
                content = content.replace(placeholder, value)
            
            # 写回文件
            file_path.write_text(content, encoding='utf-8')
        except Exception as e:
            # 忽略二进制文件或其他无法处理的文件
            pass
            
    def _rename_placeholder_files(self, directory: Path):
        """重命名包含占位符的文件和目录"""
        # 需要重命名的文件（从深到浅处理，避免路径问题）
        items_to_rename = []
        for item in directory.rglob('*'):
            if '{{COMPONENT_NAME}}' in item.name:
                items_to_rename.append(item)
        
        # 按深度排序（深的先处理）
        items_to_rename.sort(key=lambda p: len(p.parts), reverse=True)
        
        for item in items_to_rename:
            new_name = item.name.replace('{{COMPONENT_NAME}}', self.component)
            new_path = item.parent / new_name
            item.rename(new_path)
            
    def _apply_custom_enhancements(self):
        """应用定制增强（根据 AI 理解的需求）"""
        # 这里可以根据 template_data 中的信息，进一步增强组件
        # 例如：添加特定的 Props、生成更详细的 spec.md 等
        
        features = self.data.get('features', [])
        if not features:
            return
        
        # 增强 spec.md
        spec_file = self.component_dir / 'spec.md'
        if spec_file.exists():
            self._enhance_spec(spec_file, features)
        
        # 生成初始 features.json
        features_file = self.component_dir / 'features.json'
        if features_file.exists():
            self._generate_initial_features(features_file)
            
    def _enhance_spec(self, spec_file: Path, features: list):
        """增强 spec.md，添加提取的功能信息"""
        content = spec_file.read_text(encoding='utf-8')
        
        # 在 "核心功能" 部分后添加提取的功能
        features_text = "\n".join([f"- {feature}" for feature in features])
        enhancement = f"\n\n提取的核心功能:\n{features_text}\n"
        
        content = content.replace(
            "<!-- 请描述组件的核心功能 -->",
            f"<!-- 请描述组件的核心功能 -->{enhancement}"
        )
        
        spec_file.write_text(content, encoding='utf-8')
        
    def _generate_initial_features(self, features_file: Path):
        """生成初始的 features.json"""
        initial_features = [
            {
                "id": "INIT-001",
                "category": "setup",
                "priority": "high",
                "title": "组件基础结构",
                "description": "已从模板创建组件骨架",
                "status": "completed",
                "completed_at": datetime.now().isoformat(),
                "estimated_complexity": 1
            }
        ]
        
        features_file.write_text(json.dumps(initial_features, indent=2, ensure_ascii=False))


def main():
    parser = argparse.ArgumentParser(description="组件脚手架生成器（基于模板）")
    parser.add_argument("--component", required=True, help="组件名称")
    parser.add_argument("--mode", required=True, choices=["minimal", "standard", "custom"], help="生成模式")
    parser.add_argument("--template-data", required=True, help="模板数据（JSON 字符串）")
    
    args = parser.parse_args()
    
    # 解析模板数据
    try:
        template_data = json.loads(args.template_data)
    except json.JSONDecodeError as e:
        print(f"❌ 模板数据格式错误: {e}")
        sys.exit(1)
    
    # 生成脚手架
    scaffolder = ComponentScaffolder(args.component, args.mode, template_data)
    scaffolder.generate()


if __name__ == "__main__":
    main()

File: scripts/scaffold-component.js
import fs from 'node:fs/promises';
import path from 'node:path';
import { fileURLToPath } from 'node:url';
import { parseArgs } from 'node:util';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
const PROJECT_ROOT = path.resolve(__dirname, '..');
const TEMPLATE_DIR = path.join(PROJECT_ROOT, '.ai-workflow', 'templates', 'component');

class ComponentScaffolder {
  constructor(component, mode, templateData) {
    this.component = component;
    this.mode = mode;
    this.data = templateData;
    this.componentDir = path.join(PROJECT_ROOT, 'components', component);
    this.placeholders = this._generatePlaceholders();
  }

  _generatePlaceholders() {
    const parts = this.component.split('-');
    const pascalCase = parts.map(word => word.charAt(0).toUpperCase() + word.slice(1)).join('');
    
    return {
      '{{COMPONENT_NAME}}': this.component,
      '{{COMPONENT_NAME_PASCAL}}': pascalCase,
      '{{DESCRIPTION}}': this.data.description || `${this.component} 组件`,
      '{{INIT_DATE}}': new Date().toISOString(),
    };
  }

  async generate() {
    console.log(`🏗️  从模板创建 ${this.component}`);

    // 检查模板是否存在
    try {
      await fs.access(TEMPLATE_DIR);
    } catch {
      console.error(`❌ 模板目录不存在: ${TEMPLATE_DIR}`);
      console.error(`请确保 .ai-workflow/templates/component/ 目录存在`);
      process.exit(1);
    }

    // 检查组件是否已存在
    try {
      await fs.access(this.componentDir);
      console.error(`❌ 组件已存在: ${this.component}`);
      process.exit(1);
    } catch {
      // 组件不存在，可以创建
    }

    // 1. 拷贝模板目录
    console.log(`📦 拷贝模板...`);
    await this._copyDir(TEMPLATE_DIR, this.componentDir);

    // 2. 替换所有文件中的占位符
    console.log(`✏️  替换占位符...`);
    await this._replacePlaceholdersInDirectory(this.componentDir);

    // 3. 重命名包含占位符的文件
    console.log(`📝 重命名文件...`);
    await this._renamePlaceholderFiles(this.componentDir);

    // 4. 根据 mode 进行增强
    if (this.mode === 'custom' && this.data) {
      console.log(`🎨 应用定制增强...`);
      await this._applyCustomEnhancements();
    }

    console.log(`✅ 组件脚手架创建完成`);
    console.log(`📂 位置: components/${this.component}`);
  }

  async _copyDir(src, dest) {
    await fs.mkdir(dest, { recursive: true });
    const entries = await fs.readdir(src, { withFileTypes: true });

    for (const entry of entries) {
      const srcPath = path.join(src, entry.name);
      const destPath = path.join(dest, entry.name);

      if (entry.isDirectory()) {
        await this._copyDir(srcPath, destPath);
      } else {
        await fs.copyFile(srcPath, destPath);
      }
    }
  }

  async _replacePlaceholdersInDirectory(directory) {
    const entries = await fs.readdir(directory, { withFileTypes: true });

    for (const entry of entries) {
      const fullPath = path.join(directory, entry.name);
      if (entry.isDirectory()) {
        await this._replacePlaceholdersInDirectory(fullPath);
      } else if (entry.isFile()) {
        await this._replacePlaceholdersInFile(fullPath);
      }
    }
  }

  async _replacePlaceholdersInFile(filePath) {
    try {
      const ext = path.extname(filePath).toLowerCase();
      const binaryExts = ['.png', '.jpg', '.jpeg', '.gif', '.ico', '.woff', '.woff2', '.ttf'];
      if (binaryExts.includes(ext)) return;

      let content = await fs.readFile(filePath, 'utf-8');
      
      for (const [placeholder, value] of Object.entries(this.placeholders)) {
        content = content.replaceAll(placeholder, value);
      }

      await fs.writeFile(filePath, content, 'utf-8');
    } catch (e) {
      // 忽略
    }
  }

  async _renamePlaceholderFiles(directory) {
    const entries = await fs.readdir(directory, { withFileTypes: true });
    
    for (const entry of entries) {
      const fullPath = path.join(directory, entry.name);
      if (entry.isDirectory()) {
        await this._renamePlaceholderFiles(fullPath);
      }
    }

    for (const entry of entries) {
      if (entry.name.includes('{{COMPONENT_NAME}}')) {
        const oldPath = path.join(directory, entry.name);
        const newName = entry.name.replaceAll('{{COMPONENT_NAME}}', this.component);
        const newPath = path.join(directory, newName);
        await fs.rename(oldPath, newPath);
      }
    }
  }

  async _applyCustomEnhancements() {
    const features = this.data.features || [];
    if (features.length === 0) return;

    const specFile = path.join(this.componentDir, 'spec.md');
    try {
      await fs.access(specFile);
      await this._enhanceSpec(specFile, features);
    } catch {}

    const featuresFile = path.join(this.componentDir, 'features.json');
    try {
      await fs.access(featuresFile);
      await this._generateInitialFeatures(featuresFile);
    } catch {}
  }

  async _enhanceSpec(specFile, features) {
    let content = await fs.readFile(specFile, 'utf-8');
    const featuresText = features.map(f => `- ${f}`).join('\n');
    const enhancement = `\n\n提取的核心功能:\n${featuresText}\n`;
    
    content = content.replace(
      "<!-- 请描述组件的核心功能 -->",
      `<!-- 请描述组件的核心功能 -->${enhancement}`
    );
    
    await fs.writeFile(specFile, content, 'utf-8');
  }

  async _generateInitialFeatures(featuresFile) {
    const initialFeatures = [
      {
        "id": "INIT-001",
        "category": "setup",
        "priority": "high",
        "title": "组件基础结构",
        "description": "已从模板创建组件骨架",
        "status": "completed",
        "completed_at": new Date().toISOString(),
        "estimated_complexity": 1
      }
    ];
    
    await fs.writeFile(featuresFile, JSON.stringify(initialFeatures, null, 2), 'utf-8');
  }
}

async function main() {
  const { values } = parseArgs({
    options: {
      component: { type: 'string' },
      mode: { type: 'string' },
      'template-data': { type: 'string' },
    }
  });

  if (!values.component || !values.mode || !values['template-data']) {
    console.error('用法: node scripts/scaffold-component.js --component <name> --mode <minimal|standard|custom> --template-data <json>');
    process.exit(1);
  }

  let templateData = {};
  try {
    templateData = JSON.parse(values['template-data']);
  } catch (e) {
    console.error(`❌ 模板数据格式错误: ${e}`);
    process.exit(1);
  }

  const scaffolder = new ComponentScaffolder(
    values.component, 
    values.mode, 
    templateData
  );
  await scaffolder.generate();
}

main().catch(err => {
  console.error('脚本执行出错:', err);
  process.exit(1);
});
